### PR TITLE
Refacto crypto module

### PR DESCRIPTION
**Pull request by shikanime**

### CODE REVIEW

File: apps/channel_service/lib/channel_service/endpoint/cryptography.ex [hunk 1]
```diff
@@ -3,119 +3,225 @@ defmodule ChannelService.Endpoint.Cryptography do
...
 
-  @spec next(binary(), integer() | nil) :: {binary() | nil, binary()}
-  def next(raw, enc_key) do
-    case do_next(raw, enc_key) do
-      {packet, rest} -> {packet, rest}
-      nil -> {nil, raw}
+  @doc """
+  Decrypt the delimiter from a key.
+  """
+  @spec pack_delimiter(integer(), integer()) :: integer()
```
Comment (ImNotAVirus) [inline apps/channel_service/lib/channel_service/endpoint/cryptography.ex:13]: Be more precise on your typespecs for example `@type mode :: 0..3`

File: apps/channel_service/lib/channel_service/endpoint/cryptography.ex [hunk 2]
```diff
@@ -3,119 +3,225 @@ defmodule ChannelService.Endpoint.Cryptography do
...
     end
   end
 
-  @spec unpack(binary(), any()) :: packet()
-  def unpack(binary, _ \\ nil) do
-    do_unpack(binary, @table)
+  @doc """
+  Decrypt the offset from a key.
+  """
+  @spec cipher_offset(integer()) :: integer()
```
Comment (ImNotAVirus) [inline apps/channel_service/lib/channel_service/endpoint/cryptography.ex:26]: Same here, the result is 0..255

File: apps/channel_service/lib/channel_service/endpoint/cryptography.ex [hunk 3]
```diff
@@ -3,119 +3,225 @@ defmodule ChannelService.Endpoint.Cryptography do
...
+  def cipher_offset(key) do
+    band(key, 0xFF)
   end
 
   @doc """
-  Encrypt a world packet.
+  Decrypt the mode from a key.
+  """
+  @spec cipher_mode(integer()) :: integer()
```
Comment (ImNotAVirus) [inline apps/channel_service/lib/channel_service/endpoint/cryptography.ex:34]: Same

File: apps/channel_service/lib/channel_service/endpoint/cryptography.ex [hunk 4]
```diff
@@ -3,119 +3,225 @@ defmodule ChannelService.Endpoint.Cryptography do
...
-  @spec encrypt(String.t()) :: binary
-  def encrypt(packet) do
-    bytes =
-      packet
-      |> :unicode.characters_to_list(:utf8)
-      |> Enum.with_index()
+  @spec next(binary(), map()) :: {binary(), binary()}
+  def next(raw, assigns) do
+    raw
+    |> :binary.split([<<assigns.delimiter>>])
```
Comment (ImNotAVirus) [inline apps/channel_service/lib/channel_service/endpoint/cryptography.ex:50]:
```suggestion
    |> :binary.split(delimiter)
```

File: apps/channel_service/lib/channel_service/endpoint/cryptography.ex [hunk 5]
```diff
@@ -3,119 +3,225 @@ defmodule ChannelService.Endpoint.Cryptography do
...
+      {<<198, 228, 203, 145, 70, 205, 214, 220, 208, 217, 208, 196, 7, 212, 73>>, <<208, 203, 222, 209, 215, 208, 210, 218, 193, 112, 67, 220, 208, 210, 63, 199, 228, 203, 161, 16, 72, 215, 214, 221, 200, 214, 200, 214, 248, 193, 160, 65, 218, 193, 224, 66, 241, 205, 199, 228, 203, 161, 16, 72, 215, 214, 221, 200, 214, 200, 214, 248, 193, 160, 65, 218, 193, 224, 66, 241, 205>>}
   """
-  @spec encrypt(String.t()) :: binary
-  def encrypt(packet) do
-    bytes =
-      packet
-      |> :unicode.characters_to_list(:utf8)
-      |> Enum.with_index()
+  @spec next(binary(), map()) :: {binary(), binary()}
+  def next(raw, assigns) do
```
Comment (ImNotAVirus) [inline apps/channel_service/lib/channel_service/endpoint/cryptography.ex:48]: don't use assigns just for a single variable

File: apps/channel_service/lib/channel_service/endpoint/cryptography.ex [hunk 6]
```diff
@@ -3,119 +3,225 @@ defmodule ChannelService.Endpoint.Cryptography do
...
+    raw
+    |> :binary.split([<<assigns.delimiter>>])
+    |> List.to_tuple()
```
Comment (ImNotAVirus) [inline apps/channel_service/lib/channel_service/endpoint/cryptography.ex:51]: If your binary can't be splitted, the result will be a tuple with only one element

File: apps/channel_service/lib/channel_service/endpoint/cryptography.ex [hunk 7]
```diff
@@ -3,119 +3,225 @@ defmodule ChannelService.Endpoint.Cryptography do
...
+  @doc """
+  Unpack a world packet.
+
+  ## Examples
+
+      iex> ChannelService.Endpoint.Cryptography.unpack(<<135, 141, 107, 177, 64>>)
+      "49277 0"
+  """
+  @spec unpack(binary(), binary()) :: binary()
+  def unpack(binary, acc \\ <<>>)
```
Comment (ImNotAVirus) [inline apps/channel_service/lib/channel_service/endpoint/cryptography.ex:80]: Don't use a default argument here or `unpack/2` will be public. Use a `do_unpack/2` instead

File: apps/channel_service/lib/channel_service/endpoint/cryptography.ex [hunk 8]
```diff
@@ -3,119 +3,225 @@ defmodule ChannelService.Endpoint.Cryptography do
...
+
+  def unpack(<<>>, acc) do
+    acc
+  end
+
+  def unpack(<<flag, rest::binary>>, acc) do
+    if 0x7A > flag do
```
Comment (ImNotAVirus) [inline apps/channel_service/lib/channel_service/endpoint/cryptography.ex:87]: Can be done in guards

File: apps/channel_service/lib/channel_service/endpoint/cryptography.ex [hunk 9]
```diff
@@ -3,119 +3,225 @@ defmodule ChannelService.Endpoint.Cryptography do
...
+    {data, :binary.part(pack, {len, byte_size(pack) - len})}
   end
 
-  ## Private functions
+  defp bsr_permutation(c) do
+    h = bsr(c, 4)
 
-  defp do_next(raw, enc_key, acc \\ [])
-  defp do_next(<<>>, _enc_key, _acc), do: nil
+    if h != 0 and h != 0xF do
```
Comment (ImNotAVirus) [inline apps/channel_service/lib/channel_service/endpoint/cryptography.ex:105]: Guard

File: apps/channel_service/lib/channel_service/endpoint/cryptography.ex (hunk 1 above)
Comment (shikanime) [reply apps/channel_service/lib/channel_service/endpoint/cryptography.ex:13]: Forgot we could do that 👍

File: apps/channel_service/lib/channel_service/endpoint/cryptography.ex (hunk 2 above)
Comment (shikanime) [reply apps/channel_service/lib/channel_service/endpoint/cryptography.ex:26]: Forgot we could do that 👍

File: apps/channel_service/lib/channel_service/endpoint/cryptography.ex (hunk 3 above)
Comment (shikanime) [reply apps/channel_service/lib/channel_service/endpoint/cryptography.ex:34]: Forgot we could do that 👍

File: apps/channel_service/lib/channel_service/endpoint/cryptography.ex (hunk 4 above)
Comment (shikanime) [reply apps/channel_service/lib/channel_service/endpoint/cryptography.ex:50]: Integer is not a valid binary pattern

File: apps/channel_service/lib/channel_service/endpoint/cryptography.ex (hunk 9 above)
Comment (shikanime) [reply apps/channel_service/lib/channel_service/endpoint/cryptography.ex:105]: I just wanted to share the var instead, how would you have done it?

Comment (ImNotAVirus) [reply apps/channel_service/lib/channel_service/endpoint/cryptography.ex:110]: Nvm

**Comment by shikanime** [comment]
outdated

**Pull request closed without merging**